[design] Functions for explicitly fetched signature algorithms (#22672)
openssl/openssl: doc/designs/functions-for-explicitly-fetched-signature-algorithms.md

```

Added stream functionality:

``` C
int EVP_PKEY_sign_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
                        unsigned char *sig, size_t *siglen, size_t sigsize);
```

### For verifying with `EVP_SIGNATURE`

With verifying, recent development has shown a need to set the signature to
be verified against separately.  This proposal reflects that.

New initializers:

``` C
int EVP_PKEY_verify_init_ex2(EVP_PKEY_CTX *pctx,
                             EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
int EVP_PKEY_verify_recover_init_ex2(EVP_PKEY_CTX *pctx,
                                     EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
```

New signature setter:

``` C
int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
                               unsigned char *sig, size_t *siglen, size_t sigsize);
```

Added stream functionality:

``` C
int EVP_PKEY_verify_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
int EVP_PKEY_verify_final(EVP_PKEY_CTX *ctx);

int EVP_PKEY_verify_recover_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);

> int EVP_PKEY_verify_recover_final(EVP_PKEY_CTX *ctx);

Reviewer: This one needs the `unsigned char *rout, size_t *routlen` parameters.

Author: Fixed

```

Requirements on the providers
-----------------------------

Because it's not immediately obvious from a composite algorithm name what
key type it requires / supports, at least in code, allowing the use of an
explicitly fetched implementation of a composite algorithm requires that
providers cooperate by declaring what key type is required / supported by
each algorithm.

For non-composite operation algorithms (like "RSA"), this is not necessary,
see the fallback strategies below.

There are two ways this could be implemented:

1.  through an added provider function that would work like keymgmt's
    `query_operation_name` function, but would return a key type name
    instead:

    ``` C
    # define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
    OSSL_CORE_MAKE_FUNC(const char *, signature_query_key_type, (void))
    ```

> 2.  through a gettable `OSSL_PARAM`, using the param identity "keytype"

Reviewer: This returns just one type but presumably RSA-PSS-WITH-SHA256 could use both RSA and RSA-PSS keys.

Author: Eh gods, that.  How I regret making RSA-PSS a separate key type...

Author: I'll have a think about that

Reviewer 2: fyi -- I have been working on at adding RSA-OAEP as a new key type in #22344

Author: I don't like it, but I'll reduce this part to have only the first option, and let it return `const char **`.

Functions for explicitly fetched PKEY algorithms
================================================

Quick background
----------------

There are several proposed designs that end up revolving around the same
basic need, explicitly fetched signature algorithms.  The following method
type is affected by this document:

- `EVP_SIGNATURE`

Public API - Add variants of `EVP_PKEY_CTX` functionality
---------------------------------------------------------

Through OTC discussions, it's been determined that the most suitable APIs to
touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
associated functions.  They can be extended to accept an explicitly fetched
algorithm of the right type, and to be able to process infinite amounts of
data if the fetched algorithm permits it (typically, algorithms like ED25519

> or RSA-SHA256).

Reviewer: As @bernd-edlinger and @davidben wrote in this https://github.com/openssl/openssl/pull/23240#issuecomment-1916968985 and following comments this is potentially confusing to users and we might want to separate the API calls for _signing a message digest_ from calls for _signing a message_.

On the other hand the EVP_DigestSign set of API calls are misdesigned and ugly so my proposal is to:

1. Add an alias EVP_PKEY_sign_digest() to the existing EVP_PKEY_sign() call, and EVP_PKEY_verify_digest() to the existing EVP_PKEY_verify() call.
2. Add functions EVP_PKEY_sign_message(), EVP_PKEY_verify_message() that would be implemented by these ED25519, RSA-SHA256 or similar algorithms.

Author: Depending on how you read it, *message* is quite confusing too.  If you read PKCS#1 carefully, you'll note that RSASP1 and RSAVP1 also take a (limited size) message, *represented* as an integer 0 → n - 1 (they talk about "message representative", which I read to mean that the integer is a representation of the message, a limited size octet string that was passed to O2ISP, or that will get passed through I2OSP).

Author: Thinking a bit about it, it could be that *primitive* and *data* provide a better distinction, with the semantics that *primitive* only allow a one-shot style of operation, while *data* allows a stream style of operation.  So we'd end up with:

- `EVP_PKEY_sign_primitive_init()` and `EVP_PKEY_sign_primitive()` (with the old `EVP_PKEY_sign_init()` and `EVP_PKEY_sign()` as aliases)
- `EVP_PKEY_sign_data_init()`, `EVP_PKEY_sign_data_update()` and `EVP_PKEY_sign_data_final()`

... and similar for any other stuff we want to give the same treatment.

Reviewer: Yeah, that works for me.

Reviewer: Now we need to think about the provider side interface.

Reviewer: Could we just map the new APIs to the digest_sign_init()/digest_verify_init() provider functions with adding appropriate aliases through macros and add just new calls as needed?

Author: I've changed things around a bit, hopefully legibly

It must be made clear that the added functionality can not be used to
compose an algorithm from different parts.  For example, it's not possible
to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
other input that's limited to the modulus size of the RSA pkey.

### For signing with `EVP_SIGNATURE`

New initializers:

``` C
int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *pctx,
                           EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
```

Added stream functionality:

``` C
int EVP_PKEY_sign_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);

> int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,

Reviewer: These would be named EVP_PKEY_sign_data_update and EVP_PKEY_sign_data_final in my proposal.

Author: You have changed your mind along the way...  `message` → `data`

Reviewer: Yeah, I was kind of undecided whether message or data is better - your choice. :grin:

                        unsigned char *sig, size_t *siglen, size_t sigsize);
```

### For verifying with `EVP_SIGNATURE`

With verifying, recent development has shown a need to set the signature to
be verified against separately.  This proposal reflects that.

New initializers:

``` C
int EVP_PKEY_verify_init_ex2(EVP_PKEY_CTX *pctx,
                             EVP_SIGNATURE *algo, const OSSL_PARAM params[]);
int EVP_PKEY_verify_recover_init_ex2(EVP_PKEY_CTX *pctx,

>                                      EVP_SIGNATURE *algo, const OSSL_PARAM params[]);

Reviewer: I am not so sure about usefulness of verify_recover for composite signature algorithms. IMO this is pretty much RSA specific thing.

```

New signature setter:

``` C
int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
                               unsigned char *sig, size_t *siglen, size_t sigsize);
```

Added stream functionality:

``` C
int EVP_PKEY_verify_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);

> int EVP_PKEY_verify_final(EVP_PKEY_CTX *ctx);

Reviewer: Similarly to above these would be verify_data_update and verify_data_final.

int EVP_PKEY_verify_recover_update(EVP_PKEY_CTX *ctx, unsigned char *in, size_t *inlen);
int EVP_PKEY_verify_recover_final(EVP_PKEY_CTX *ctx,

>                                   unsigned char *rout, size_t *routlen);

Reviewer: I do not think the verify_recover should have a streaming variant.

Author: Hmm, ok

```

Requirements on the providers
-----------------------------

Because it's not immediately obvious from a composite algorithm name what
key type it requires / supports, at least in code, allowing the use of an
explicitly fetched implementation of a composite algorithm requires that
providers cooperate by declaring what key type is required / supported by
each algorithm.

For non-composite operation algorithms (like "RSA"), this is not necessary,
see the fallback strategies below.

There are two ways this could be implemented:

1.  through an added provider function that would work like keymgmt's
    `query_operation_name` function, but would return a key type name
    instead:

    ``` C
    # define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
    OSSL_CORE_MAKE_FUNC(const char *, signature_query_key_type, (void))
    ```

2.  through a gettable `OSSL_PARAM`, using the param identity "keytype"

Furthermore, the public API above requires added provider functionality:

``` C
# define OSSL_FUNC_SIGNATURE_SIGN_UPDATE            26
# define OSSL_FUNC_SIGNATURE_SIGN_FINAL             27
OSSL_CORE_MAKE_FUNC(int, signature_sign_update, (void *ctx,  unsigned char *sig,

>                                                  size_t *siglen, size_t sigsize,

Reviewer 3: Should output params be last?

"primitive" functions.  This design essentially proposes an extension to
also allow streaming functionality through an *update* and a *final*
function.  Discussions have revealed that it is potentially confusing to
confound "primitive" functionality with "streaming" functionality into the
same name, so this design also includes a naming proposal that distinguishes
the two uses.

### Making it possible to verify with an early signature

There are newer verifying algorithms that need to receive the signature
before processing the data.  This is particularly important for streaming
functionality.  This design proposes a mechanism to accomodate this, which
is different than the mechanism you get with `EVP_PKEY_verify()`.

Public API - API Reference
--------------------------

### For signing with `EVP_SIGNATURE`

#### Primitive / Limit size oneshot

``` C
int EVP_PKEY_sign_primitive_init(EVP_PKEY_CTX *pctx,
                                 EVP_SIGNATURE *algo,
                                 const OSSL_PARAM params[]);
int EVP_PKEY_sign_primitive(EVP_PKEY_CTX *ctx,
                            unsigned char *sig, size_t *siglen, size_t sigsize,

>                             const unsigned char *tbs, size_t tbslen);

Reviewer: I assumed this will be just an alias to EVP_PKEY_sign().

Author: Implementation detail. But yeah

```

#### Streaming

``` C
int EVP_PKEY_sign_data_init(EVP_PKEY_CTX *pctx, EVP_SIGNATURE *algo,
                            const OSSL_PARAM params[]);

> int EVP_PKEY_sign_data_update(EVP_PKEY_CTX *ctx,

Reviewer: IMO we should have also an EVP_PKEY_sign_data() one shot function with a respective provider call that could be possibly implemented as combined update/final call for providers that would not bother implementing the one shot call.

Author: Sure.  Would you think it's acceptable that the provider *only* implements the oneshot function (i.e. not update or final)?

                                           const OSSL_PARAM params[]);
/* Oneshot functions */
int EVP_PKEY_verify_primitive(EVP_PKEY_CTX *pctx,
                              const unsigned char *sig, size_t siglen,
                              const unsigned char *tbs, size_t tbslen);
int EVP_PKEY_verify_recover_primitive(EVP_PKEY_CTX *pctx,
                                      unsigned char *rout, size_t *routlen,
                                      size_t routsize,
                                      cnnst unsigned char *sig, size_t siglen);
```

#### Streaming

Do note that there's no proposed streaming definition for verify recover
functionality.

``` C
/* Initializers */
int EVP_PKEY_verify_data_init(EVP_PKEY_CTX *pctx, EVP_SIGNATURE *algo,
                              const OSSL_PARAM params[]);
/* Signature setter */
int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,
                               unsigned char *sig, size_t siglen,
                               size_t sigsize);
/* Update and final */
int EVP_PKEY_verify_data_update(EVP_PKEY_CTX *ctx,
                                const unsigned char *in, size_t inlen);

> int EVP_PKEY_verify_data_final(EVP_PKEY_CTX *ctx);

Reviewer: Similarly to signature I'd propose having also EVP_PKEY_verify_data() oneshot call.

```

> 

Reviewer: Nit: extra empty line.

like keymgmt's `query_operation_name` function, but would return a NULL
terminated array of key type name instead:

``` C
# define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
OSSL_CORE_MAKE_FUNC(const char **, signature_query_key_type, (void))
```

Furthermore, the public API above requires added provider functionality:

``` C
# define OSSL_FUNC_SIGNATURE_SIGN_UPDATE            26
# define OSSL_FUNC_SIGNATURE_SIGN_FINAL             27
OSSL_CORE_MAKE_FUNC(int, signature_sign_update, (void *ctx,
                                                 const unsigned char *in,
                                                 size_t inlen))
OSSL_CORE_MAKE_FUNC(int, signature_sign_final, (void *ctx,  unsigned char *sig,
                                                size_t *siglen, size_t sigsize))

# define OSSL_FUNC_SIGNATURE_VERIFY_UPDATE          28
# define OSSL_FUNC_SIGNATURE_VERIFY_FINAL           29
OSSL_CORE_MAKE_FUNC(int, signature_verify_update,
                    (void *ctx, const unsigned char *in, size_t inlen))
/*
 * signature_verify_final requires that the signature to be verified against
 * is specified via an OSSL_PARAM.
 */

> OSSL_CORE_MAKE_FUNC(int, signature_verify_final, (void *ctx))

Reviewer: I assume that we need to reflect the EVP api calls in the providers interface as well?

Author: No, please no. As a provider author, I *protest* this proliferation of a gazillion functions that do the same essential thing

Reviewer: I do not understand how we can implement the proposed EVP calls in terms of the existing provider calls.

Author: The only additional calls I can see making sense would be `signature_sign_data_init` and `signature_verify_data_init`.  If that was all you meant, then sure.

But for example, I don't see why we'd need to have two different oneshot functions.

Reviewer: I was actually thinking about not adding additional _init functions because the existing ones could cover both sign_data and sign_primitive. But then you of course need two different oneshot functions because otherwise you cannot distinguish whether you want the primitive operation or not.

Reviewer: If we have a different init operation for the primitive and data calls then the oneshot op can be the same, that's true.

Author: Actually...  I don't see much of a need to have different inits, even.  `primitive` vs `data` is really semantic sugar for the application, 'cause in the end, what differentiates the "primitive" vs "data" functionality is that the former requires calling `EVP_PKEY_CTX_set_signature_md()` while the latter doesn't (for "data", it should even be prohibited - not settable - since the "data" algorithms are expected to be composite algorithms, i.e. hash function implied).

In terms of oneshot, the difference between "primitive" and "data" is really that the former might error with some sort of error indicating that the input (tbs) is too large.

The flip side with having alternative function indexes (one for "primitive" and one for "data") is the possible implication that you should be able to have both.  That's a source of unnecessary error and confusion for provider authors.

Author: In the end, there are a number of possible indicators to see if an `EVP_SIGNATURE` can be considered "primitive", which can be checked in libcrypto:

- The presence or not of `OSSL_FUNC_SIGNATURE_SIGN_UPDATE` and `OSSL_FUNC_SIGNATURE_VERIFY_UPDATE`.
- The presence or not of `OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE`
- If there is a `OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE`, does its response match the key type of the given (through `EVP_PKEY_CTX`) `EVP_PKEY`?

Reviewer: I am afraid this will be very confusing to provider authors.

Author: With my provider author hat on, I disagree.  I find the multitude of diverse functions to implement a much bigger pile of confusion...  and error prone outside of the provider's control (hard question: did libcrypto call the right variant of, say, update, for this use case?)

- `EVP_SIGNATURE`

Public API - Add variants of `EVP_PKEY_CTX` functionality
---------------------------------------------------------

Through OTC discussions, it's been determined that the most suitable APIs to
touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
`EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
associated functions.  They can be extended to accept an explicitly fetched
algorithm of the right type, and to be able to process infinite amounts of
data if the fetched algorithm permits it (typically, algorithms like ED25519
or RSA-SHA256).

It must be made clear that the added functionality can not be used to
compose an algorithm from different parts.  For example, it's not possible
to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
other input that's limited to the modulus size of the RSA pkey.

### Making things less confusing with distinct function names

So far, `EVP_PKEY_sign()` and friends are only expected to act as

> "primitive" functions.  This design essentially proposes an extension to

Reviewer 4: I think we should not be using `primitve function` here. I think this paragraph  should try to untangle the confusion that comes from mixing up two distinct classes here:
   - API type which is a _one-shot_ vs. _streaming_ (IUF as @vdukhovni noted)
   - operation type which is either a cryptographic primitive itself (particular digest/encryption/sign/verify alg. such as AES or SHA256) or a compound operation which is made of more than one cryptographic primitive (such as digest + sign) 

also putting `primitive function` to google will get me to calculus which is even more confusing. So I would not be stealing terms from math here.

Author: I'm actually stealing "primitive" from [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#section-5) (`EVP_PKEY_sign` (etc) functions [actually, the provider] does give direct access to them, for RSA, under the condition that the flag `RSA_NO_PADDING` is set and that no hash function has been declared with `EVP_PKEY_CTX_set_signature_md()`, and with the [I2OSP and OSP2I](https://www.rfc-editor.org/rfc/rfc8017#section-4) primitives implied).

I'm wrong when I'm saying "only", though, since also message hashes are supported by our providers (and are properly encoded/decoded when the correct flags are set, and the signature hash is specified).

Author: I've expanded on this text to clarify the current functionality a bit more

also allow streaming functionality through an *update* and a *final*
function.  Discussions have revealed that it is potentially confusing to
confound pure "primitive" functionality with functionality that expects a
digest as input, as well as with "streaming" functionality into the same
name, so this design also includes a naming proposal that distinguishes
the two uses.

### Making it possible to verify with an early signature

There are newer verifying algorithms that need to receive the signature
before processing the data.  This is particularly important for streaming
functionality.  This design proposes a mechanism to accomodate this, which
is different than the mechanism you get with `EVP_PKEY_verify()`.

Public API - API Reference
--------------------------

### For signing with `EVP_SIGNATURE`

#### Signing a digest

``` C
int EVP_PKEY_sign_init_for_digest(EVP_PKEY_CTX *pctx,
                                  EVP_SIGNATURE *algo,
                                  const OSSL_PARAM params[]);
```

(notably, a fetched "RSA" should probably not accept `RSA_NO_PADDING` when
initialized with this call)

#### Signing a stream

``` C

> int EVP_PKEY_sign_init_for_message(EVP_PKEY_CTX *pctx,

Reviewer: I would call it EVP_PKEY_sign_message_init().

                                   EVP_SIGNATURE *algo,
                                   const OSSL_PARAM params[]);
int EVP_PKEY_sign_update(EVP_PKEY_CTX *ctx,
                         const unsigned char *in, size_t inlen);
int EVP_PKEY_sign_final(EVP_PKEY_CTX *ctx,
                        unsigned char *sig, size_t *siglen);

> ```

Reviewer: We will also need a oneshot. I.e., EVP_PKEY_sign_message(). And I would change those two above to EVP_PKEY_sign_message_update() and EVP_PKEY_sign_message_final().

application.

[^1]: the term "primitive" is borrowed from [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#section-5)

### Making it possible to verify with an early signature

There are newer verifying algorithms that need to receive the signature
before processing the data.  This is particularly important for streaming
functionality.  This design proposes a mechanism to accomodate this for the
streaming style functionality.

New public API - API Reference
------------------------------

### For limited input size / oneshot signing with `EVP_SIGNATURE`

``` C
int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *pctx,
                           EVP_SIGNATURE *algo,
                           const OSSL_PARAM params[]);
```

### For signing a stream with `EVP_SIGNATURE`

``` C
int EVP_PKEY_sign_message_init(EVP_PKEY_CTX *pctx,
                               EVP_SIGNATURE *algo,
                               const OSSL_PARAM params[]);
int EVP_PKEY_sign_message_update(EVP_PKEY_CTX *ctx,
                                 const unsigned char *in,
                                 size_t inlen);
int EVP_PKEY_sign_message_final(EVP_PKEY_CTX *ctx,
                                unsigned char *sig,

>                                 size_t *siglen);

Reviewer: What about the one-shot `EVP_PKEY_sign_message()` function? Please note that for example ED25519 requires two passes over the message -> this has to be a one-shot operation. Yeah, you could say that you can call EVP_PKEY_sign_message_update only once for ED25519 but that sounds like a kludge.

Author: EVP_PKEY_sign() can be used just as well.  You did propose an alias macro, something I'm mentioning higher up.  Do you actually want me to spell that out here?

Reviewer: Yeah, please add it for clarity.

Functions for explicitly fetched PKEY algorithms
================================================

Quick background
----------------

There are several proposed designs that end up revolving around the same
basic need, explicitly fetched signature algorithms.  The following method
type is affected by this document:

- `EVP_SIGNATURE`

Public API - Add variants of `EVP_PKEY_CTX` functionality
---------------------------------------------------------

> Through OTC discussions, it's been determined that the most suitable APIs to
> touch is the collection of `EVP_PKEY_` functions.  In this case, it involves
> `EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()` with
> associated functions.  They can be extended to accept an explicitly fetched
> algorithm of the right type, and to be able to process infinite amounts of
> data if the fetched algorithm permits it (typically, algorithms like ED25519
> or RSA-SHA256).

Reviewer 5: ```suggestion
Through OTC discussions, it's been determined that the most suitable APIs to
touch are the `EVP_PKEY_` functions.
Specifically, `EVP_PKEY_sign()`, `EVP_PKEY_verify()`, `EVP_PKEY_verify_recover()`
and related functions.
They can be extended to accept an explicitly fetched algorithm of the right type,
and to be able to incrementally process indefinite length data streams when the
fetched algorithm permits it (typically, algorithms like ED25519 or RSA-SHA256).
```
Does Ed25519 really support streaming? Or are you thinking of Ed25519ph?
It was my impression that pure Ed25519 required the entire message in memory
and did not support IUF-style (Init, Update, Final) signing.

Author: Ah, I should remove ED25519 from there, 'cause indeed, the entire length must be known from the start, which had escaped me when I wrote this passage.

> It must be made clear that the added functionality can not be used to
> compose an algorithm from different parts.  For example, it's not possible
> to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
> specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
> `EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
> other input that's limited to the modulus size of the RSA pkey.

Reviewer 5: Editorial nit aside, I find this paragraph confusing. Please explain what you mean more clearly.
```suggestion
It must be made clear that the added functionality cannot be used to
compose an algorithm from different parts.  For example, it's not possible
to specify a `EVP_SIGNATURE` "RSA" and combine it with a parameter that
specifies the hash "SHA256" to get the "RSA-SHA256" functionality.  For an
`EVP_SIGNATURE` "RSA", the input is still expected to be a digest, or some
other input that's limited to the modulus size of the RSA pkey.
```

### Making things less confusing with distinct function names

> So far, `EVP_PKEY_sign()` and friends are only expected to act on the

Reviewer 5: ```suggestion
Till now, `EVP_PKEY_sign()` and friends are only expected to act on the
```

pre-computed digest of a message (under the condition that proper flags
and signature md are specified using functions like
`EVP_PKEY_CTX_set_rsa_padding()` and `EVP_PKEY_CTX_set_signature_md()`),
or to act as "primitive" [^1] functions (under the condition that proper
flags are specified, like `RSA_NO_PADDING` for RSA signatures).

This design proposes an extension to also allow full (not pre-hashed)
messages to be passed, in a streaming style through an *update* and a
*final* function.

> Discussions have revealed that it is potentially confusing to confound the

Reviewer 5: ```suggestion
Discussions have revealed that it is potentially confusing to conflate the
```

current functionality with streaming style functionality into the same name,
so this design separates those out with specific init / update / final
functions for that purpose.  For oneshot functionality, `EVP_PKEY_sign()`

> and `EVP_PKEY_verify()` remain supported, through an alias for the
> application.

Reviewer 5: ```suggestion
and `EVP_PKEY_verify()` remain supported.
```
If you feel that the mechanism by which they remain supported should be explained, the explanation should be less terse. It is not clear what "an alias for the application" meant.

[^1]: the term "primitive" is borrowed from [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#section-5)

### Making it possible to verify with an early signature

> There are newer verifying algorithms that need to receive the signature
> before processing the data.  This is particularly important for streaming
> functionality.  This design proposes a mechanism to accomodate this for the
> streaming style functionality.

Reviewer 5: ```suggestion
Some more recent verification algorithms need to obtain the signature
before processing the data.
This is particularly important for streaming modes of operation.
This design proposes a mechanism to accomodate these algorithms
and modes of operation.
```

                           const OSSL_PARAM params[]);
```

### For signing a stream with `EVP_SIGNATURE`

``` C
int EVP_PKEY_sign_message_init(EVP_PKEY_CTX *pctx,
                               EVP_SIGNATURE *algo,
                               const OSSL_PARAM params[]);
int EVP_PKEY_sign_message_update(EVP_PKEY_CTX *ctx,
                                 const unsigned char *in,
                                 size_t inlen);
int EVP_PKEY_sign_message_final(EVP_PKEY_CTX *ctx,
                                unsigned char *sig,
                                size_t *siglen);
#define EVP_PKEY_sign_message(ctx,sig,siglen,tbs,tbslen) \
    EVP_PKEY_sign(ctx,sig,siglen,tbs,tbslen)
```

### For limited input size / oneshot verification with `EVP_SIGNATURE`

``` C
int EVP_PKEY_verify_init_ex2(EVP_PKEY_CTX *pctx,
                             EVP_SIGNATURE *algo,
                             const OSSL_PARAM params[]);
```

### For verifying a stream with `EVP_SIGNATURE`

``` C
/* Initializers */
int EVP_PKEY_verify_message_init(EVP_PKEY_CTX *pctx,
                                 EVP_SIGNATURE *algo,
                                 const OSSL_PARAM params[]);
/* Signature setter */

> int EVP_PKEY_CTX_set_signature(EVP_PKEY_CTX *pctx,

Reviewer 3: Is this name ok?
It doesnt match all the others here (the CTX bit) - should we keep it consistent with the others..

Why doesnt the name have _verify_ in its name if it is only used for verification?

Having a seperate function is good for the case where the signature is needed early. For the other case it causes an unnecessary mem copy since the signature cant be applied via the message_final().
Is this ok?

Reviewer: What about calling the function EVP_PKEY_CTX_set0_signature() or EVP_PKEY_verify_message_set0_signature()?

And use octet ref instead of octet string param. Of course that means usage in applications for the case where the message needs to be set before calling the update function(s) will be more complicated. Maybe the extra allocation/memcpy()/free call is not really relevant given the complexity of the signature verification.

Author: > Is the EVP_PKEY_CTX_set_signature name ok?
> It doesnt match all the others here (the CTX bit) - should we keep it consistent with the others, even if it is wrong.

I modelled that after already existing patterns, see EVP_PKEY_CTX_set_signature_md() and EVP_PKEY_CTX_set_rsa_padding().

> Why doesnt the name have _verify_ in its name if it is only used for verification?

Because the *thing* that's being set is the signature to verify against.  We don't have a *thing* called "verification", do we?

> Having a separate function is good for the exceptional case where the signature is needed early. For the normal case it causes an unnecessary memcopy since the signature cant be applied via the message_final(). Is this ok?

This was a choice to avoid causing confusion by allowing two ways to set that signature.  What happens if someone gives the signature to verify against using both EVP_PKEY_CTX_set_signature() and EVP_PKEY_verify_message_final() (if we imagine the latter with the added arguments to set the signature that way)?

Author: > What about calling the function EVP_PKEY_CTX_set0_signature() or EVP_PKEY_verify_message_set0_signature()?

I could go for either.

> And use octet ref instead of octet string param.

Which means making it another provider function...  well, considering that a signature is mandatory for verification, that makes sense.

Reviewer 3: I prefer EVP_PKEY_verify_message_set0_signature() since it matches the other (wrong) ones :)

Author: set0?  set1?

                               unsigned char *sig, size_t siglen,
                               size_t sigsize);
/* Update and final */
int EVP_PKEY_verify_message_update(EVP_PKEY_CTX *ctx,
                                   const unsigned char *in,
                                   size_t inlen);
int EVP_PKEY_verify_message_final(EVP_PKEY_CTX *ctx);

#define EVP_PKEY_verify_message(ctx,sig,siglen,tbs,tbslen) \

>     EVP_PKEY_sign(ctx,sig,siglen,tbs,tbslen)

Reviewer 5: This seems to be a cut/paste error from the signature side.

Author: Yes it was

```

### For verify_recover with `EVP_SIGNATURE`

> According to feedback, a stream style interface is uninteresting for
> verify_recover, so we only specify a new init function.

Reviewer 5: ```suggestion
Preliminary feedback suggests that a streaming interface is uninteresting for
verify_recover, so we only specify a new init function.
```

``` C
/* Initializers */
int EVP_PKEY_verify_recover_init_ex2(EVP_PKEY_CTX *pctx,
                                     EVP_SIGNATURE *algo,
                                     const OSSL_PARAM params[]);
```

Requirements on the providers
-----------------------------

> Because it's not immediately obvious from a composite algorithm name what
> key type it requires / supports, at least in code, allowing the use of an

Reviewer 5: ```suggestion
Because it isn't always obvious from a composite algorithm name what
key type it requires / supports, allowing the use of an
```
When you say "key type", what do you mean?
- Public vs. Private?
- Particular asymmetric algorithm?
- Distinctions like RSA unrestricted vs. RSA PSS?
- Something else?

Author: We use the term "key type" in other places for "RSA", "DSA", "EC"...

I a provider context, these are essentially keymgmt names.

Reviewer 5: > We use the term "key type" in other places for "RSA", "DSA", "EC"...
> 
> I a provider context, these are essentially keymgmt names.

So, essentially the value of EVP_PKEY_type()? An in-context example could help avoid confusion in the minds of the uninitiated, perhaps: key type ("RSA", "EC", ...).
EVP_PKEY_type()

There are newer verifying algorithms that need to receive the signature
before processing the data.  This is particularly important for streaming
functionality.  This design proposes a mechanism to accomodate this for the
streaming style functionality.

New public API - API Reference
------------------------------

### For limited input size / oneshot signing with `EVP_SIGNATURE`

``` C
int EVP_PKEY_sign_init_ex2(EVP_PKEY_CTX *pctx,
                           EVP_SIGNATURE *algo,
                           const OSSL_PARAM params[]);
```

### For signing a stream with `EVP_SIGNATURE`

``` C
int EVP_PKEY_sign_message_init(EVP_PKEY_CTX *pctx,
                               EVP_SIGNATURE *algo,
                               const OSSL_PARAM params[]);
int EVP_PKEY_sign_message_update(EVP_PKEY_CTX *ctx,
                                 const unsigned char *in,
                                 size_t inlen);
int EVP_PKEY_sign_message_final(EVP_PKEY_CTX *ctx,
                                unsigned char *sig,
                                size_t *siglen);
```

### For limited input size / oneshot verification with `EVP_SIGNATURE`

``` C

> int EVP_PKEY_verify_ex2(EVP_PKEY_CTX *pctx,

Reviewer: I think you meant `EVP_PKEY_verify_init_ex2` here?

Author: Oops.  Yes
